### PR TITLE
Fix viewport with transparent bg changed to solid black if enable fxaa or debanding.

### PR DESCRIPTION
**Pull request by Kinwailo**
If enable fxaa or debanding in viewport with transparent bg, the transparent bg will change to solid black color. It is because the shader always output alpha as 1.0.

*Bugsquad edit: This closes https://github.com/godotengine/godot/issues/45208.*

### CODE REVIEW

**Comment by Calinou** [comment]
Does this fully address https://github.com/godotengine/godot/issues/45208, or are there still parts of the tonemapping shader that don't play well with Transparent Bg?

**Comment by Kinwailo** [comment]
> Good catch slightly_smiling_face
> 
> Does this fully address #45208, or are there still parts of the tonemapping shader that don't play well with Transparent Bg?

Post processing is bypass if transparent is enabled in driver code, and the shader of glow / dof produce "black edge". I have changed the code and shader, should fix #45208.

**Comment by Calinou** [comment]
I tested this PR rebased against the latest `3.x` branch as of writing. It indeed fixes the issue in the [original MRP](https://github.com/godotengine/godot/files/5820414/PostBug.zip) :slightly_smiling_face: 

### Before



### After (with this PR)



When Transparent Bg is disabled, the appearance is identical:

### Before



### After (with this PR)

File: drivers/gles3/shaders/tonemap.glsl [hunk 1]
```diff
@@ -286,7 +286,17 @@ vec3 apply_glow(vec3 color, vec3 glow) { // apply glow using the selected blendi
...
+#endif
+
+#ifndef USE_GLOW_SOFTLIGHT // softlight has no effect on black color
+	// compute the alpha from glow
+	float a = max(max(glow.r, glow.g), glow.b);
+	color.a = a + color.a * (1 - a);
+	if (color.a == 0.0)
+		color.rgb = vec3(0.0);
+	else if (color.a < 1.0)
+		color.rgb /= color.a;
```
Comment (Calinou) [inline drivers/gles3/shaders/tonemap.glsl:299]: Use braces for single-line `if` statements, as per the code style guidelines:

```suggestion
	if (color.a == 0.0) {
		color.rgb = vec3(0.0);
	} else if (color.a < 1.0) {
		color.rgb /= color.a;
	}
```

File: drivers/gles2/shaders/tonemap.glsl [hunk 2]
```diff
@@ -198,7 +198,17 @@ vec3 apply_glow(vec3 color, vec3 glow) { // apply glow using the selected blendi
...
+#endif
+
+#ifndef USE_GLOW_SOFTLIGHT // softlight has no effect on black color
+	// compute the alpha from glow
+	float a = max(max(glow.r, glow.g), glow.b);
+	color.a = a + color.a * (1 - a);
+	if (color.a == 0.0)
+		color.rgb = vec3(0.0);
+	else if (color.a < 1.0)
+		color.rgb /= color.a;
```
Comment (Calinou) [inline drivers/gles2/shaders/tonemap.glsl:211]:
```suggestion
	if (color.a == 0.0) {
		color.rgb = vec3(0.0);
	} else if (color.a < 1.0) {
		color.rgb /= color.a;
	}
```

**Review by Calinou — CHANGES_REQUESTED** [review]

**Comment by bfloch** [comment]
I was able to verify this with a project that was suffering from the issue in 3.4.3. Works great, no complaints.

**Comment by bfloch** [comment]
A negative side-effect I just noticed: In transparent viewports that are setup via a ViewportTexture in a TextureRect and not ViewportController the (supposedly transparent) background looks overly bright.
Somehow I noticed that transparency does not seem to work in either 3.4.3 or this PR this way, but that may be a separate issue or something wrong on my  end.

The screenshot shows the PR at the top:
godot_transparency_effect

**Comment by Calinou** [comment]
Looks good to me, please squash the commits together so we can merge this. See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase) for instructions :slightly_smiling_face:

**Comment by clayjohn** [comment]
This looks pretty good and is very promising. 

The only issue I noticed while testing is that FXAA doesn't weight the transparent samples by their alpha, so you end up with black edges when using FXAA.

This shouldn't be too hard to fix, it may just be a matter of premultiplying the sampled colour by its transparency so that transparent pixels don't contribute their color. 

_FXAA on_
Screenshot from 2022-05-04 13-18-14

_FXAA off_
Screenshot from 2022-05-04 13-18-18

File: drivers/gles2/shaders/tonemap.glsl [hunk 3]
```diff
@@ -253,19 +264,21 @@ vec3 apply_fxaa(vec3 color, vec2 uv_interp, vec2 pixel_size) {
...
-	} else {
-		return rgbB;
+	float lumaB = dot(rgbB.rgb, luma) - ((1 - rgbB.a) / 8.0);
+	vec4 output = ((lumaB < lumaMin) || (lumaB > lumaMax)) ? rgbA : rgbB;
+	if (output.a == 0.0) {
+		output.rgb = vec3(0.0);
+	} else if (output.a < 1.0) {
+		output.rgb /= output.a;
 	}
+	return output;
```
Comment (clayjohn) [inline drivers/gles2/shaders/tonemap.glsl:277]: ``output`` is a reserved word in glsl so this shader will fail to compile when drivers strictly enforce the glsl specification (typically Intel drivers). To compile, the name just needs to be changed. 

After making this change everything seems to work well!

```suggestion
	vec4 output = ((lumaB < lumaMin) || (lumaB > lumaMax)) ? rgbA : rgbB;
	if (output.a == 0.0) {
		output.rgb = vec3(0.0);
	} else if (output.a < 1.0) {
		output.rgb /= output.a;
	}
	return output;
```

**Review by clayjohn — CHANGES_REQUESTED** [review]
Just the one change needed. Other than that, this is looking good to go!

**Comment by clayjohn** [comment]
> > Just the one change needed. Other than that, this is looking good to go!
> 
> should change the shader for gles3 too?

Yes please

**Comment by Kinwailo** [comment]
> > > Just the one change needed. Other than that, this is looking good to go!
> > 
> > 
> > should change the shader for gles3 too?
> 
> Yes please

changed

**Review by clayjohn — APPROVED** [review]
Looks great! Thank you!

**Review by Calinou — APPROVED** [review]
Thanks! Congratulations for your first merged pull request :tada:

**Pull request merged**